File: desktop/tts.go
package main

import (
	"bufio"
	"context"
	"encoding/json"
	"fmt"
	"os/exec"
	"strings"
	"sync"
	"syscall"

	"github.com/wailsapp/wails/v2/pkg/runtime"
)

// ttsState constants emitted via the "tts-state" event.
const (
	ttsStatePlaying = "playing"
	ttsStatePaused  = "paused"
	ttsStateStopped = "stopped"
	ttsStateError   = "error"
)

type ttsSession struct {
	cmd    *exec.Cmd
	cancel context.CancelFunc
	mu     sync.Mutex
	state  string // playing, paused, stopped
}

type ttsChunk struct {
	Type  string `json:"type"`            // "audio", "done", "error"
	Audio string `json:"audio,omitempty"` // base64-encoded WAV data
	Error string `json:"error,omitempty"`
}

// ttsScript is the Python one-liner that streams Kokoro output as JSON lines.
// Each chunk is a JSON object: {"type":"audio","audio":"<base64>"} or
// {"type":"progress","percent":50} or {"type":"done"}.
const ttsScript = `
import sys, json, base64, io
try:
    import soundfile as sf
    from kokoro import KPipeline
except ImportError as e:
    print(json.dumps({"type":"error","error":"Missing dependency: "+str(e)+". Install with: pip install kokoro soundfile"}))
    sys.exit(1)

text = sys.argv[1]
voice = sys.argv[2]
speed = float(sys.argv[3])

try:
    pipeline = KPipeline(lang_code=voice[0])
    for gs, ps, audio in pipeline(text, voice=voice, speed=speed):
        buf = io.BytesIO()
        sf.write(buf, audio, 24000, format="WAV")
        b64 = base64.b64encode(buf.getvalue()).decode()
        print(json.dumps({"type":"audio","audio":b64}))
        sys.stdout.flush()
    print(json.dumps({"type":"done","percent":100}))
except Exception as e:
    print(json.dumps({"type":"error","error":str(e)}))
    sys.exit(1)
`

// StartTTS begins text-to-speech synthesis, reading voice and speed from
// the user's saved settings. It delegates to startTTS which spawns the
// Kokoro subprocess.
func (a *App) StartTTS(text string) error {
	a.settingsMu.Lock()
	s := a.loadSettingsLocked()
	a.settingsMu.Unlock()

	voice := s.TTSVoice
	speed := s.TTSSpeed
	return a.startTTS(text, voice, speed)
}

// startTTS begins text-to-speech synthesis. It spawns a Python subprocess
// running Kokoro, streaming audio chunks and progress back to the frontend
// via Wails events. Only one TTS session is active at a time; calling
// startTTS while one is running stops the previous session first.
func (a *App) startTTS(text string, voice string, speed float64) error {
	if strings.TrimSpace(text) == "" {
		return fmt.Errorf("text is empty")
	}
	if voice == "" {
		voice = "af_heart"
	}
	if speed <= 0 {
		speed = 1.0
	}

	// Stop any existing session
	a.stopTTSLocked()

	ctx, cancel := context.WithCancel(context.Background())

	cmd := exec.CommandContext(ctx, "python3", "-c", ttsScript, text, voice, fmt.Sprintf("%.2f", speed))

	stdout, err := cmd.StdoutPipe()
	if err != nil {
		cancel()
		return fmt.Errorf("create stdout pipe: %w", err)
	}
	cmd.Stderr = nil // ignore stderr; errors come as JSON on stdout

	if err := cmd.Start(); err != nil {
		cancel()
		return fmt.Errorf("start tts process: %w", err)
	}

	sess := &ttsSession{
		cmd:    cmd,
		cancel: cancel,
		state:  ttsStatePlaying,
	}

	a.ttsMu.Lock()
	a.ttsSession = sess
	a.ttsMu.Unlock()

	runtime.EventsEmit(a.ctx, "tts-state", ttsStatePlaying)

	// Read goroutine: parse JSON lines from the Python process
	go func() {
		scanner := bufio.NewScanner(stdout)
		// Allow large audio chunks (up to 10MB per line)
		scanner.Buffer(make([]byte, 0, 64*1024), 10*1024*1024)

		for scanner.Scan() {
			var chunk ttsChunk
			if err := json.Unmarshal(scanner.Bytes(), &chunk); err != nil {
				continue
			}

			switch chunk.Type {
			case "audio":
				runtime.EventsEmit(a.ctx, "tts-audio", chunk.Audio)
			case "done":
				// synthesis complete; playback may still be ongoing
			case "error":
				runtime.EventsEmit(a.ctx, "tts-state", ttsStateError)
				runtime.EventsEmit(a.ctx, "tts-error", chunk.Error)
			}
		}

		// Wait for process exit
		_ = cmd.Wait()

		// Mark session stopped unless already cleaned up
		a.ttsMu.Lock()
		if a.ttsSession == sess {
			a.ttsSession = nil
		}
		a.ttsMu.Unlock()

		sess.mu.Lock()
		wasPlaying := sess.state == ttsStatePlaying
		sess.state = ttsStateStopped
		sess.mu.Unlock()

		if wasPlaying {
			// Normal completion
			runtime.EventsEmit(a.ctx, "tts-state", ttsStateStopped)
		}
	}()

	return nil
}

// StopTTS terminates the current TTS session.
func (a *App) StopTTS() {
	a.stopTTSLocked()
}

// stopTTSLocked stops the active TTS session if one exists.
func (a *App) stopTTSLocked() {
	a.ttsMu.Lock()
	sess := a.ttsSession
	a.ttsSession = nil
	a.ttsMu.Unlock()

	if sess == nil {
		return
	}

	sess.mu.Lock()
	sess.state = ttsStateStopped
	sess.mu.Unlock()

	// Resume if paused so the process can be killed cleanly
	if sess.cmd.Process != nil {
		_ = sess.cmd.Process.Signal(syscall.SIGCONT)
	}
	sess.cancel()

	runtime.EventsEmit(a.ctx, "tts-state", ttsStateStopped)
}

// PauseTTS pauses the current TTS session by sending SIGSTOP to the
// Python subprocess.
func (a *App) PauseTTS() error {
	a.ttsMu.Lock()
	sess := a.ttsSession
	a.ttsMu.Unlock()

	if sess == nil {
		return fmt.Errorf("no active tts session")
	}

	sess.mu.Lock()
	defer sess.mu.Unlock()

	if sess.state != ttsStatePlaying {
		return fmt.Errorf("tts is not playing (state: %s)", sess.state)
	}

	if sess.cmd.Process != nil {
		if err := sess.cmd.Process.Signal(syscall.SIGSTOP); err != nil {
			return fmt.Errorf("pause tts: %w", err)
		}
	}
	sess.state = ttsStatePaused
	runtime.EventsEmit(a.ctx, "tts-state", ttsStatePaused)
	return nil
}

// ResumeTTS resumes a paused TTS session by sending SIGCONT to the
// Python subprocess.
func (a *App) ResumeTTS() error {
	a.ttsMu.Lock()
	sess := a.ttsSession
	a.ttsMu.Unlock()

	if sess == nil {
		return fmt.Errorf("no active tts session")
	}

	sess.mu.Lock()
	defer sess.mu.Unlock()

	if sess.state != ttsStatePaused {
		return fmt.Errorf("tts is not paused (state: %s)", sess.state)
	}

	if sess.cmd.Process != nil {
		if err := sess.cmd.Process.Signal(syscall.SIGCONT); err != nil {
			return fmt.Errorf("resume tts: %w", err)
		}
	}
	sess.state = ttsStatePlaying
	runtime.EventsEmit(a.ctx, "tts-state", ttsStatePlaying)
	return nil
}

func (a *App) CheckKokoroInstalled() bool {
	cmd := exec.Command("python3", "-c", "from kokoro import KPipeline; import soundfile")
	return cmd.Run() == nil
}

// InstallKokoro installs the kokoro and soundfile Python packages.
func (a *App) InstallKokoro() error {
	cmd := exec.Command("pip3", "install", "kokoro", "soundfile")
	out, err := cmd.CombinedOutput()
	if err != nil {
		return fmt.Errorf("pip3 install failed: %s\n%s", err, string(out))
	}
	return nil
}

func (a *App) UninstallKokoro() error {
	cmd := exec.Command("pip3", "uninstall", "-y", "kokoro", "soundfile")
	out, err := cmd.CombinedOutput()
	if err != nil {
		return fmt.Errorf("pip3 uninstall failed: %s\n%s", err, string(out))
	}
	return nil
}

File: desktop/frontend/src-tauri/src/tts.rs
// Kokoro text-to-speech — port of desktop/tts.go.
//
// One session at a time: python3 runs an inline Kokoro script that prints
// {"type":"audio","audio":"<base64 WAV>"} JSON lines; a reader thread re-emits
// each as a "tts-audio" event (bare base64 string — a complete 24kHz WAV the
// frontend's Web Audio player decodes per chunk). State transitions go out as
// "tts-state" ("playing"/"paused"/"stopped"/"error"); errors as "tts-error".
// Pause/Resume are SIGSTOP/SIGCONT; Stop wakes (SIGCONT) then kills.
use crate::config;
use std::io::BufRead;
use std::process::{Child, Command, Stdio};
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::{Arc, Mutex};
use tauri::{AppHandle, Emitter, Manager, State};

const PLAYING: &str = "playing";
const PAUSED: &str = "paused";
const STOPPED: &str = "stopped";

// Verbatim from desktop/tts.go (leading + trailing newline preserved). r#"..."#
// avoids escaping the embedded JSON double-quotes.
const TTS_SCRIPT: &str = r#"
import sys, json, base64, io
try:
    import soundfile as sf
    from kokoro import KPipeline
except ImportError as e:
    print(json.dumps({"type":"error","error":"Missing dependency: "+str(e)+". Install with: pip install kokoro soundfile"}))
    sys.exit(1)

text = sys.argv[1]
voice = sys.argv[2]
speed = float(sys.argv[3])

try:
    pipeline = KPipeline(lang_code=voice[0])
    for gs, ps, audio in pipeline(text, voice=voice, speed=speed):
        buf = io.BytesIO()
        sf.write(buf, audio, 24000, format="WAV")
        b64 = base64.b64encode(buf.getvalue()).decode()
        print(json.dumps({"type":"audio","audio":b64}))
        sys.stdout.flush()
    print(json.dumps({"type":"done","percent":100}))
except Exception as e:
    print(json.dumps({"type":"error","error":str(e)}))
    sys.exit(1)
"#;

#[derive(serde::Deserialize)]
struct TtsChunk {
    #[serde(rename = "type")]
    kind: String,
    #[serde(default)]
    audio: String,
    #[serde(default)]
    error: String,
}

struct TtsSession {
    child: Child,
    pid: i32,
    state: String,
    id: u64, // generation guard: only the current session's reader emits "stopped"
}

pub struct TtsState {
    inner: Arc<Mutex<Option<TtsSession>>>,
    counter: AtomicU64,
}

impl Default for TtsState {
    fn default() -> Self {
        Self {
            inner: Arc::new(Mutex::new(None)),
            counter: AtomicU64::new(0),
        }
    }
}

type Inner = Arc<Mutex<Option<TtsSession>>>;

#[tauri::command(async)]
pub fn start_tts(app: AppHandle, state: State<'_, TtsState>, text: String) -> Result<(), String> {
    if text.trim().is_empty() {
        return Err("text is empty".into());
    }
    let s = config::load_settings();
    let mut voice = s
        .get("ttsVoice")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();
    let mut speed = s.get("ttsSpeed").and_then(|v| v.as_f64()).unwrap_or(0.0);
    if voice.is_empty() {
        voice = "af_heart".into();
    }
    if speed <= 0.0 {
        speed = 1.0;
    }

    stop_internal(&app, &state.inner); // tear down any prior session first

    let mut child = Command::new("python3")
        .arg("-c")
        .arg(TTS_SCRIPT)
        .arg(&text) // sys.argv[1]
        .arg(&voice) // sys.argv[2]
        .arg(format!("{speed:.2}")) // sys.argv[3]
        .stdin(Stdio::null())
        .stdout(Stdio::piped())
        .stderr(Stdio::null()) // errors arrive as JSON on stdout
        .spawn()
        .map_err(|e| format!("start tts process: {e}"))?;
    let stdout = child.stdout.take().ok_or("create stdout pipe")?;
    let pid = child.id() as i32;
    let id = state.counter.fetch_add(1, Ordering::SeqCst) + 1;
    *state.inner.lock().unwrap() = Some(TtsSession {
        child,
        pid,
        state: PLAYING.into(),
        id,
    });
    let _ = app.emit("tts-state", PLAYING);

    let app2 = app.clone();
    let inner: Inner = state.inner.clone();
    std::thread::spawn(move || {
        // BufRead::lines grows unbounded, so base64 WAV chunks aren't truncated
        // (Go needed an explicit 10 MB scanner buffer; here it's automatic).
        for line in std::io::BufReader::new(stdout)
            .lines()
            .map_while(Result::ok)
        {
            let Ok(chunk) = serde_json::from_str::<TtsChunk>(&line) else {
                continue;
            };
            match chunk.kind.as_str() {
                "audio" => {
                    let _ = app2.emit("tts-audio", chunk.audio);
                }
                "done" => {} // no-op (frontend ends via the audio player)
                "error" => {
                    let _ = app2.emit("tts-state", "error");
                    let _ = app2.emit("tts-error", chunk.error);
                }
                _ => {}
            }
        }
        // EOF: if this is still the current session, reap it and emit "stopped"
        // (iff it was playing). A Stop / new Start already took it -> do nothing.
        let mut guard = inner.lock().unwrap();
        let is_current = guard.as_ref().map(|s| s.id == id).unwrap_or(false);
        if is_current {
            let mut sess = guard.take().unwrap();
            drop(guard);
            let was_playing = sess.state == PLAYING;
            let _ = sess.child.wait();
            if was_playing {
                let _ = app2.emit("tts-state", STOPPED);
            }
        }
    });
    Ok(())
}

#[tauri::command(async)]
pub fn stop_tts(app: AppHandle, state: State<'_, TtsState>) {
    stop_internal(&app, &state.inner);
}

fn stop_internal(app: &AppHandle, inner: &Inner) {
    let sess = inner.lock().unwrap().take();
    let Some(mut sess) = sess else {
        return;
    };
    // Wake a possibly-paused process so the kill isn't queued behind SIGSTOP,
    // then terminate. SIGKILL (== Go's context-cancel) dies promptly.
    unsafe {
        libc::kill(sess.pid, libc::SIGCONT);
        libc::kill(sess.pid, libc::SIGKILL);
    }
    let _ = sess.child.wait(); // reap; the reader thread will see None -> no double emit
    let _ = app.emit("tts-state", STOPPED);
}

#[tauri::command(async)]
pub fn pause_tts(app: AppHandle, state: State<'_, TtsState>) -> Result<(), String> {
    let mut guard = state.inner.lock().unwrap();
    let sess = guard.as_mut().ok_or("no active tts session")?;
    if sess.state != PLAYING {
        return Err(format!("tts is not playing (state: {})", sess.state));
    }
    if unsafe { libc::kill(sess.pid, libc::SIGSTOP) } != 0 {
        return Err(format!("pause tts: {}", std::io::Error::last_os_error()));
    }
    sess.state = PAUSED.into();
    drop(guard);
    let _ = app.emit("tts-state", PAUSED);
    Ok(())
}

#[tauri::command(async)]
pub fn resume_tts(app: AppHandle, state: State<'_, TtsState>) -> Result<(), String> {
    let mut guard = state.inner.lock().unwrap();
    let sess = guard.as_mut().ok_or("no active tts session")?;
    if sess.state != PAUSED {
        return Err(format!("tts is not paused (state: {})", sess.state));
    }
    if unsafe { libc::kill(sess.pid, libc::SIGCONT) } != 0 {
        return Err(format!("resume tts: {}", std::io::Error::last_os_error()));
    }
    sess.state = PLAYING.into();
    drop(guard);
    let _ = app.emit("tts-state", PLAYING);
    Ok(())
}

#[tauri::command(async)]
pub fn check_kokoro_installed() -> bool {
    Command::new("python3")
        .args(["-c", "from kokoro import KPipeline; import soundfile"])
        .status()
        .map(|s| s.success())
        .unwrap_or(false)
}

#[tauri::command(async)]
pub fn install_kokoro() -> Result<(), String> {
    let out = Command::new("pip3")
        .args(["install", "kokoro", "soundfile"])
        .output()
        .map_err(|e| e.to_string())?;
    if !out.status.success() {
        return Err(format!(
            "pip3 install failed: {}\n{}",
            out.status,
            combined(&out)
        ));
    }
    Ok(())
}

#[tauri::command(async)]
pub fn uninstall_kokoro() -> Result<(), String> {
    let out = Command::new("pip3")
        .args(["uninstall", "-y", "kokoro", "soundfile"])
        .output()
        .map_err(|e| e.to_string())?;
    if !out.status.success() {
        return Err(format!(
            "pip3 uninstall failed: {}\n{}",
            out.status,
            combined(&out)
        ));
    }
    Ok(())
}

fn combined(out: &std::process::Output) -> String {
    let mut s = String::from_utf8_lossy(&out.stdout).into_owned();
    s.push_str(&String::from_utf8_lossy(&out.stderr));
    s
}

/// Kill any live TTS child on app exit (prevents an orphaned suspended python).
pub fn stop_on_exit(app: &AppHandle) {
    let state = app.state::<TtsState>();
    stop_internal(app, &state.inner);
}
